### PR TITLE
http_client: reduce number of packets sent for small bodies

### PR DESCRIPTION
**Pull request by jeffro256**
Right now, HTTP headers and the HTTP body are sent separately. This PR change sends the bodies and headers combined when appropriate. For small bodies, this can reduce the number of packets sent by 1. For a series of HTTP requests made with small bodies, the number of round trips can be reduced by up to 50%. 

You can test this by analyzing HTTP traffic with a packet sniffer like Wireshark.

### CODE REVIEW

**Comment by jeffro256** [comment]
To test:

1. Tack the following code to the end of the file `tests/unit_tests/http.cpp` for the the master branch:
```
TEST(HTTP_Client, Simple_Post)
{
  http::http_simple_client client;
  client.set_server("52.86.68.46", "80", boost::none, epee::net_utils::ssl_support_t::e_ssl_support_disabled);
  
  const http::http_response_info* presponse_info = NULL;
  EXPECT_TRUE(client.invoke("/post", "POST", "meep mop", std::chrono::seconds(10), &presponse_info));
}

```
2. Compile branch
3. Start capturing packets with Wireshark
4. Run `./build/Linux/<BRANCH NAME>/release/tests/unit_tests/unit_tests --gtest_filter=HTTP_Client.*`
5. Stop capturing with Wireshark
5. Count how many packets it takes to send request headers and body
6. Repeat steps 1-5 for this PR branch

**Comment by plowsof** [comment]
i have an issue compiling:
```
monero/tests/unit_tests/http.cpp: In member function ‘virtual void HTTP_Client_Simple_Post_Test::TestBody()’:
/home/human/dev/monero-gui/monero/tests/unit_tests/http.cpp:749:9: error: ‘http_simple_client’ is not a member of ‘{anonymous}::http’
  749 |   http::http_simple_client client;
      |         ^~~~~~~~~~~~~~~~~~
monero/tests/unit_tests/http.cpp:750:3: error: ‘client’ was not declared in this scope
  750 |   client.set_server("52.86.68.46", "80", boost::none, epee::net_utils::ssl_support_t::e_ssl_support_disabled);
      |   ^~~~~~
monero/tests/unit_tests/http.cpp:750:72: error: ‘epee::net_utils::ssl_support_t’ has not been declared
  750 |   client.set_server("52.86.68.46", "80", boost::none, epee::net_utils::ssl_support_t::e_ssl_support_disabled);
      |                 
```

File: contrib/epee/include/net/http_client.h [hunk 1]
```diff
@@ -209,21 +219,29 @@ namespace net_utils
...
 					req_buff += "\r\n";
 					//--
 
-					bool res = m_net_client.send(req_buff, timeout);
-					CHECK_AND_ASSERT_MES(res, false, "HTTP_CLIENT: Failed to SEND");
-					if(body.size())
+					if (do_copy_body) // small body
+					{
+						// Copy headers + body together and potentially send one fewer packet
+						req_buff.append(body.data(), body.size());
```
Comment (iamamyth) [inline contrib/epee/include/net/http_client.h:233]: Why not change the client to offer a writev-style method? Then you can skip the copies.

File: contrib/epee/include/net/net_helper.h [hunk 2]
```diff
@@ -487,79 +415,6 @@ namespace net_utils
...
-				LOG_ERROR("Some problems at read, message: " << er.what());
-        m_connected = false;
-				return false;
-			}
-			catch(...)
-			{
-				LOG_ERROR("Some fatal problems at read.");
-				return false;
-			}
-
```
Comment (iamamyth) [inline contrib/epee/include/net/net_helper.h:557]: Might be easier to split the dead code piece into its own PR.

**Comment by jeffro256** [comment]
@plowsof Sorry I forgot that you also have to `#include "net/http_client.h"`

File: contrib/epee/include/net/http_client.h (hunk 1 above)
Comment (jeffro256) [reply contrib/epee/include/net/http_client.h:233]: `writev` isn't portable, would require much more code changes, and probably wouldn't be worth the effort since we only have a maximum of two distinct buffers to send. Since we're only copying small buffers in this PR, I'm making the performance bet that the time spent copying this small buffer is dwarfed by the network time it takes for TCP to make a round trip.

File: contrib/epee/include/net/net_helper.h (hunk 2 above)
Comment (jeffro256) [reply contrib/epee/include/net/net_helper.h:557]: You're probably right.

Comment (jeffro256) [reply contrib/epee/include/net/net_helper.h]: Done

**Comment by plowsof** [comment]
with this PR:
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from HTTP_Client
[ RUN      ] HTTP_Client.Simple_Post
[       OK ] HTTP_Client.Simple_Post (208 ms)
[----------] 1 test from HTTP_Client (208 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (208 ms total)
[  PASSED  ] 1 test.
```
Screenshot from 2023-03-19 23-48-55


without:
```
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from HTTP_Client
[ RUN      ] HTTP_Client.Simple_Post
[       OK ] HTTP_Client.Simple_Post (301 ms)
[----------] 1 test from HTTP_Client (301 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (301 ms total)
[  PASSED  ] 1 test.
```
Screenshot from 2023-03-20 00-14-34

Is this correct? i have `ip.addr == 52.86.68.46` as a filter in wireshark

**Comment by jeffro256** [comment]
Yup! That looks just like how mine looked. Wireshark doesn't recognize that some of the TCP packets are being used for HTTP as well, but it you look through the contents of packets without the PR, you can see a packet with just the headers and a packet with the string "meep mop".

**Comment by jeffro256** [comment]
So we went from sending/receiving 14 packets for this interaction down to 10.

**Comment by 0xFFFC0000** [comment]
I read the changes. To me this looks a great little optimization. Thanks @jeffro256 

I am running the tests in a few minutes and will update this comment.

**Review by 0xFFFC0000 — APPROVED** [review]
LGTM. Passed the tests.

File: contrib/epee/include/net/http_client.h (hunk 1 above)
Comment (vtnerd) [reply contrib/epee/include/net/http_client.h:233]: I think @iamamyth was suggesting a change to the `m_net_client` interface to accept N buffers. I haven't looked too deeply on this, the easiest is probably `span<const boost::string_ref>` where a temporary array gets created here on the stack.
